x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java

```diff
@@ -53,26 +140,122 @@ public String getType() {
+        public InferenceProcessor create(Map<String, Processor.Factory> processorFactories, String tag, Map<String, Object> config)
             throws Exception {
+
+            if (this.maxIngestProcessors <= currentInferenceProcessors) {
+                throw new ElasticsearchStatusException("Max number of inference processors reached, total inference processors [{}]. " +
+                    "Adjust the setting [{}]: [{}] if a greater number is desired.",
+                    RestStatus.CONFLICT,
+                    currentInferenceProcessors,
+                    MAX_INFERENCE_PROCESSORS.getKey(),
+                    maxIngestProcessors);
+            }
+
             String modelId = ConfigurationUtils.readStringProperty(TYPE, tag, config, MODEL_ID);
-            return new InferenceProcessor(client, tag, modelId);
+            String targetField = ConfigurationUtils.readStringProperty(TYPE, tag, config, TARGET_FIELD);
+            Map<String, String> fieldMapping = ConfigurationUtils.readOptionalMap(TYPE, tag, config, FIELD_MAPPINGS);
+            InferenceConfig inferenceConfig = inferenceConfigFromMap(ConfigurationUtils.readMap(TYPE, tag, config, INFERENCE_CONFIG));
+            String modelInfoField = ConfigurationUtils.readStringProperty(TYPE, tag, config, MODEL_INFO_FIELD, "_model_info");
+            if (modelInfoField != null && tag != null) {
+                modelInfoField += "." + tag;
+            }
+            return new InferenceProcessor(client, tag, targetField, modelId, inferenceConfig, fieldMapping, modelInfoField);
         }
 
-        @Override
-        public void accept(ClusterState clusterState) {
+        // Package private for testing
+        void setMaxIngestProcessors(int maxIngestProcessors) {
+            logger.trace("updating setting maxIngestProcessors from [{}] to [{}]", this.maxIngestProcessors, maxIngestProcessors);
+            this.maxIngestProcessors = maxIngestProcessors;
+        }
+
+        InferenceConfig inferenceConfigFromMap(Map<String, Object> inferenceConfig) throws IOException {
+            ExceptionsHelper.requireNonNull(inferenceConfig, INFERENCE_CONFIG);
+
+            if (inferenceConfig.keySet().size() != 1) {
+                throw ExceptionsHelper.badRequestException("{} must be an object with one inference type mapped to an object.",
+                    INFERENCE_CONFIG);
+            }
+            Object value = inferenceConfig.values().iterator().next();
+
+            if ((value instanceof Map<?, ?>) == false) {
+                throw ExceptionsHelper.badRequestException("{} must be an object with one inference type mapped to an object.",
+                    INFERENCE_CONFIG);
+            }
+            @SuppressWarnings("unchecked")
+            Map<String, Object> valueMap = (Map<String, Object>)value;
+
+            if (inferenceConfig.containsKey(ClassificationConfig.NAME)) {
```

Comment:
I'm wondering how many places in the code will need to be changed when if we introduce a third inference type. Maybe not a problem now, but something to keep in mind if we decide the type based on a string here.

Reply:
what sucks is that the inference processor objects are not `toXContent` objects. So, I cannot use the existing named objects.